src/uuid.cr

```diff
@@ -176,6 +206,116 @@ struct UUID
+      end
+    else
+      Random::Secure.random_bytes(uuid.to_slice[10..15])
+    end
+
+    new(uuid, version: UUID::Version::V1, variant: UUID::Variant::RFC4122)
+  end
+
+  # Generates RFC 4122 v2 UUID.
+  #
+  # Version 2 UUIDs are generated using the current time, the local machine’s MAC address,
+  # and the local user or group ID. However, they are not widely used due to their limitations.
+  # For a given domain/id pair, the same token may be returned for a duration of up to 7 minutes
+  # and 10 seconds.
+  #
+  # * `domain`: The `Domain` of the uuid.
+  # * `id`: The id depends on the domain, for the `Domain::Person` usually the local user id
+  #    is used, for `Domain::Group` usually the local group id is used.
+  # * `node_id`: In case the traditional MAC address based approach should be taken the
+  #   `node_id` can be provided. Otherwise secure random is used.
+  def self.v2(domain : Domain, id : UInt32, node_id : MAC? = nil) : self
+    uuid = v1(node_id: node_id).bytes
+    uuid[6] = (uuid[6] & 0x0f) | 0x20 # Version 2
+    uuid[9] = domain.to_u8
+    IO::ByteFormat::BigEndian.encode(id, uuid.to_slice[0..3])
+    new(uuid, version: UUID::Version::V2, variant: UUID::Variant::RFC4122)
+  end
+
+  # Generates RFC 4122 v3 UUID.
+  #
+  # * `name`: The name used to generate the UUID, it can be a string of any size.
+  # * `namespace`: Specifies the type of the name, usually one of `Namespace`
+  def self.v3(name : String, namespace : UUID) : self
+    hash = Digest::MD5.digest do |ctx|
+      ctx.update namespace.bytes
+      ctx.update name
+    end
+    new(hash[0...16], version: UUID::Version::V3, variant: UUID::Variant::RFC4122)
+  end
+
+  # Generates RFC 4122 v4 UUID.
+  #
+  # It is strongly recommended to use a cryptographically random source for
+  # *random*, such as `Random::Secure`.
+  def self.v4(random r : Random = Random::Secure) : self
+    random(r)
+  end
+
+  # Generates RFC 4122 v5 UUID.
+  #
+  # * `name`: The name used to generate the UUID, it can be a string of any size.
+  # * `namespace`: Specifies the type of the name, usually one of `Namespace`
+  def self.v5(name : String, namespace : UUID) : self
+    hash = Digest::SHA1.digest do |ctx|
+      ctx.update namespace.bytes
+      ctx.update name
+    end
+    new(hash[0...16], version: UUID::Version::V5, variant: UUID::Variant::RFC4122)
+  end
+
+  {% for name in %w(DNS URL OID X500) %}
+    # Generates RFC 4122 v3 UUID with the `Namespace::{{ name.id }}`.
+    #
+    # * `name`: The name used to generate the UUID, it can be a string of any size.
+    def self.v3_{{ name.id.downcase }}(name : String)
+      v3(name, Namespace::{{ name.id }})
+    end
+
+    # Generates RFC 4122 v5 UUID with the `Namespace::{{ name.id }}`.
+    #
+    # * `name`: The name used to generate the UUID, it can be a string of any size.
+    def self.v5_{{ name.id.downcase }}(name : String)
+      v5(name, Namespace::{{ name.id }})
+    end
+  {% end %}
```

Comment:
Optional suggestion:

```suggestion
  {% for name in %w(DNS URL OID X500).map(&.id) %}
    # Generates RFC 4122 v3 UUID with the `Namespace::{{ name }}`.
    #
    # * `name`: The name used to generate the UUID, it can be a string of any size.
    def self.v3_{{ name.downcase }}(name : String)
      v3(name, Namespace::{{ name }})
    end

    # Generates RFC 4122 v5 UUID with the `Namespace::{{ name }}`.
    #
    # * `name`: The name used to generate the UUID, it can be a string of any size.
    def self.v5_{{ name.downcase }}(name : String)
      v5(name, Namespace::{{ name }})
    end
  {% end %}
```

Reply:
Thanks for your review, I updated the code as you suggested.